### PR TITLE
CI: Use MSYS2 for Windows build

### PR DESCRIPTION
**Pull request by eht16**
Please welcome a new way to build Geany and Geany-Plugins for Windows in the CI.

What's in here:
- the workflow job to checkout Geany and Infrastructure repositories
- use the supporting scripts in https://github.com/geany/infrastructure/tree/add_ci_builders/builders to
  - build a Docker image with a full MSYS2 installation (or pull the image from the Geany Docker registry if available)
  - use this Docker image to:
    - build Geany and create and test an installer
    - store the Geany installer as artifact on Github

Mostly all of the work is done in the builder scripts on https://github.com/geany/infrastructure/tree/add_ci_builders/builders.
Unfortunately, it grew more than than expected in terms of amount of code.

I tried to document at least the basics so anyone else than me will be available to get an idea what is happening.
Let me know if it worked :).

There is a difference to the Linux CI build which more or less aims at using a rather old setup to ensure backwards compability.
The Mingw-w64 build instead always pulls the latest available packages from the MSYS2 repositories.
I think this is OK as this is also what we do on releases and since we bundle all dependencies into the installer (for release builds as well as for CI builds), the result should be consistent.

In preparation for the upcoming GTK version requirement change, I updated the Linux build to Ubuntu 20.04 LTS (which has a recent enough GTK3 version).

Best of all: during the CI build Windows installers are created for Geany and Geany-Plugins and they are stored as artifacts and can be downloaded later on (up to 30 days).

Remarks:
- [ ] this branch needs a bit more testing with this PR
- [x] we need to merge the PR https://github.com/geany/infrastructure/pull/7 first

For now, the necessary Docker image was uploaded to https://github.com/geany/infrastructure/pkgs/container/geany-mingw64-ci by me manually. Basically we could push the image also from the CI job but I'm hesitating to allow this because of the risks it offers to malicious users. Maybe I'll add a separate job in the infrastructure repository later to build and push the image there and restrict this job to admins

### CODE REVIEW

File: scripts/ci_mingw64_geany.sh [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+
+convert_text_files_to_crlf() {
+	log "Converting line endings to CRLF in text files"
+	for text_file_pattern in ${GEANY_RELEASE_TEXTFILE_PATTERNS[@]}; do
+		for text_file in $(ls ${text_file_pattern}); do
+			mime_type=$(file --brief --mime-type ${text_file})
+			case $mime_type in text/*)
+				unix2dos ${text_file}
```
Comment (kugel-) [inline scripts/ci_mingw64_geany.sh:223]: Isn't that something that should be done in geany build itself? Anyone who builds Geany from source on Windows should do this right?

Comment (elextr) [reply scripts/ci_mingw64_geany.sh:223]: Git should convert line endings on checkout since the default value of `core.eol` is `native` but maybe MSYS2 doesn't register "native" as CRLF?  Maybe the scripts in MSYS2 should do `git config core.eol crlf`?

Comment (eht16) [reply scripts/ci_mingw64_geany.sh:223]: I don't expect from Geany's build system to know whether it need to convert text files to the target system's preferences.
IMO this is very Windows specific and basically only necessary to be able to read our text files with Windows' Notepad which is not able to handle different line endings.

In Geany, we would need to properly integrate it in Autotools and Meson and also depend on "unix2dos".

Regarding Git's `core.eol`: I'm not what happens on a native Windows clone but here we use the already cloned repository from the CI job and even if we would clone it ourselves, it would be LF as everything in this script is executed on Linux.

Changing the setting to `crlf` might have other side effects we do not want.

Comment (elextr) [reply scripts/ci_mingw64_geany.sh:223]: > Windows' Notepad which is not able to handle different line endings.

Works on my windows 11 tablet[^1].  Google says it was fixed in 2018, so any  _supported_ version of Windows should be ok ;-P

> Regarding Git's core.eol: I'm not what happens on a native Windows clone but here we use the already cloned repository from the CI job and even if we would clone it ourselves

Ok, so its necessary for the CI job, but I was really addressing @kugel- "Anyone who builds Geany from source on Windows should do this right?" but IIUC their checkout would do it for them with default git settings.  If tarballs are going to be buildable on windows I guess its needed there as well since the tarball files are LF endings unless we make a special Windows tarball ... or just say they are *x or modern Windows only.

> Changing the setting to crlf might have other side effects we do not want.

Yes, IIUC "Bad things" will happen if we merge anything back on Linux with that setting.

[^1]: No, its a _tablet_ so Geany is no use as I don't develop on it, sorry can't help with windows builds ;-P

Comment (kugel-) [reply scripts/ci_mingw64_geany.sh:223]: Well the CI builds shouldn't have "more features" than normal builds from source - source code is king!! If a hard dependency in the normal build is not wanted then there could be an optional dependency that is _coincidentally_ met by the CI.

I wouldn't want manual git or tarball builds to regress compared to CI builds (e.g. on Windows 7). Stuff like this is hard to track down even for experienced contributors.

IMO: Let's open an issue regarding the line ending issue. Then we either deal with it in the normal build or just live with it as suggested by @elextr . I can help with meson if that's a problem.

File: scripts/ci_mingw64_geany.sh [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+	log "Running autogen.sh"
+	./autogen.sh
+	log "Running configure"
+	./configure ${CONFIGURE_OPTIONS}
+	log "Running make"
+	make
+	log "Running install-strip"
+	make install-strip
+
+	cd /
```
Comment (kugel-) [inline scripts/ci_mingw64_geany.sh:184]: ????

File: scripts/ci_mingw64_geany.sh [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+		echo "Skip signing due to missing certificate"
+	fi
+}
+
+
+sign_geany_binaries() {
+	log "Signing Geany binary files"
+	for binary_file_pattern in ${GEANY_RELEASE_BINARY_PATTERNS[@]}; do
+		for binary_file in $(ls ${binary_file_pattern}); do
+			sign_file ${binary_file}
```
Comment (kugel-) [inline scripts/ci_mingw64_geany.sh:211]: Why not sign all files?

**Comment by kugel-** [comment]
Can you add a stub with an (perhaps commented out) example on how to build new dependencies for CI builds? I'm still afraid that it's too hard for a random guy's PR to add a dependency without breaking CI.

File: scripts/ci_mingw64_geany.sh (hunk 1 above)
Comment (elextr) [reply scripts/ci_mingw64_geany.sh:223]: > IMO: Let's open an issue regarding the line ending issue. 

Agreed, and I would have thought the tarball problem would exist already, can anyone test it?

Comment (eht16) [reply scripts/ci_mingw64_geany.sh:223]: Could any of you open that issue?
I still don't get the problem you see, as outlined above, a GIT clone on Windows probably gives the correct line endings and so I'm not really able to create an issue for an issue I don't understand :).

Here we are talking about a native Linux clone with native LF line endings and those we need to convert. IMO this is special to the CI build and so needs to be done here.

**Comment by eht16** [comment]
> Can you add a stub with an (perhaps commented out) example on how to build new dependencies for CI builds?

I don't get it :(. Where should I add the example and what do you mean by "build new dependencies"?

File: scripts/ci_mingw64_geany.sh (hunk 2 above)
Comment (eht16) [reply scripts/ci_mingw64_geany.sh]: Removed, thanks for the pointer.

**Comment by kugel-** [comment]
Nevermind, the G-P PR shows how to install additional build dependencies.

File: scripts/ci_mingw64_geany.sh (hunk 1 above)
Comment (kugel-) [reply scripts/ci_mingw64_geany.sh:221]: A tarball created on Linux will have wrong line endings on Windows. The normal build system should take care of that instead of being restricted to CI.

And how do you create Win32 releases if not using CI (as you mentioned in another comment)?

Comment (eht16) [reply scripts/ci_mingw64_geany.sh:221]: > A tarball created on Linux will have wrong line endings on Windows. The normal build system should take care of that instead of being restricted to CI.

As said above, if you want to do this, please create an issue or even better a PR to make the changes to the build system(s). Then we can remove the line conversion here.
I won't do it because I still think it's not such a big deal and I won't spend the time effort it needs.

> And how do you create Win32 releases if not using CI (as you mentioned in another comment)?

As described on https://wiki.geany.org/howtos/win32/msys2, especially using the [attached](https://wiki.geany.org/_media/howtos/win32/geany-release.py.txt) script which also does the line ending conversion to CRLF.
In short, using a clean GIT clone with the release tag. At least with my GIT config the relevant files all have LF line endings and therefore need to be converted.
Using a release tarball wouldn't help here because it is created on Linux and so also has LF line endings.

Comment (kugel-) [reply scripts/ci_mingw64_geany.sh:221]: I opened #3350 

I would like to see the conversion to be removed from this PR since I'm not even sure it's necessary at all. Don't recent Windows version deal fine with LF line endings. I don't know if notepad alone is an indication but see https://devblogs.microsoft.com/commandline/extended-eol-in-notepad/, this fix is included in Windows 10 and 11.

Comment (eht16) [reply scripts/ci_mingw64_geany.sh]: Thanks for creating the issue and the investigation.
Nice that Notepad now can handle non-CRLF line endings. This was the main reason I convert the line endings.

So I just removed the code from the build script here and in the G-P counterpart.

**Comment by eht16** [comment]
I just cleaned up the commit history and removed temporary test settings, so this can be merged.

**Review by kugel- — APPROVED** [review]
